android/src/main/java/com/tailscale/ipn/NetworkChangeCallback.kt

```diff
@@ -6,94 +6,159 @@ import android.net.ConnectivityManager
 import android.net.LinkProperties
 import android.net.Network
 import android.net.NetworkCapabilities
+import android.net.NetworkRequest
 import android.util.Log
 import libtailscale.Libtailscale
 import java.net.InetAddress
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.withLock
 
 object NetworkChangeCallback {
 
   private const val TAG = "NetworkChangeCallback"
 
-  // Cache LinkProperties and NetworkCapabilities since synchronous ConnectivityManager calls are
-  // prone to races.
-  // Since there is no guarantee for which update might come first, maybe update DNS configs on
-  // both.
-  val networkCapabilitiesCache = mutableMapOf<Network, NetworkCapabilities>()
-  val linkPropertiesCache = mutableMapOf<Network, LinkProperties>()
-
-  // requestDefaultNetworkCallback receives notifications about the default network. Listen for
-  // changes to the capabilities, which are guaranteed to come after a network becomes available per
-  // https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onAvailable(android.net.Network),
-  // in order to filter on non-VPN networks.
+  private data class NetworkInfo(
+    var caps: NetworkCapabilities,
+    var linkProps: LinkProperties
+  )
+
+  private val lock = ReentrantLock()
+  private val activeNetworks = mutableMapOf<Network, NetworkInfo>() // keyed by Network
+
+  // monitorDnsChanges sets up a network callback to monitor changes to the
+  // system's network state and update the DNS configuration when interfaces
+  // become available or properties of those interfaces change.
   fun monitorDnsChanges(connectivityManager: ConnectivityManager, dns: DnsConfig) {
+    val networkConnectivityRequest =
+      NetworkRequest.Builder()
+      .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+      .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
+      .build()
 
-    connectivityManager.registerDefaultNetworkCallback(
+    // Use registerNetworkCallback to listen for updates from all networks, and
+    // then update DNS configs for the best network.
+    //
+    // Note that we can't use registerDefaultNetworkCallback because the
+    // default network used by Tailscale will always show up with capability
+    // NOT_VPN=false, and we must filter out NOT_VPN networks to avoid routing
+    // loops.
+    connectivityManager.registerNetworkCallback(
+        networkConnectivityRequest,
         object : ConnectivityManager.NetworkCallback() {
+          override fun onAvailable(network: Network) {
+            super.onAvailable(network)
+
+            Log.d(TAG, "onAvailable: network ${network.toString()}")
```

Comment:
I learned that you don't need to explicitly call .toString() in string interpolation, you can just reference the object directly and Kotlin automatically calls .toString() for you

Reply:
Huh, TIL; thanks!